README
listpod-app-lite
=============

Convert YouTube Playlist to Podcast feed. branche of listpod.tv.
-------------

http://listpod.tv and http://www.youtubemp4.com are usufull for enjoying Podcasing with YouTube.

But YouTube returns "402 Payment Required", so they are not working now.

listpod-app-lite is client application instead of these.

This single script enable to convert YouTube playlist to podcast feed.

You can subscribe this feed on iTunes etc.

Enjoy your podcasting with YouTube.


USAGE
-------------

> $ listpod-app-lite -y youtube_playlist_url

OR if YouTube return 402 Payment Required on your ip address space.

> $ listpod-app-lite -y youtube_playlist_url -c safari_cookie_path( OSX only ...) -s AnyEvent


AUTHOR
-------------

Yusuke Wada <yusuke at kamawada.com>

Changes
Revision history for Perl extension ListPod::App::Lite

0.01    Sun Mar 14 23:32:53 2010
        - original version


lib/ListPod/App/Lite/RSS.pm
package ListPod::App::Lite::RSS;
use strict;
use warnings;
use Text::MicroTemplate qw(build_mt);

sub render {
    my ( $self, $args ) = @_;
    my $pos = tell DATA;
    local $/; my $t = <DATA>;
    seek DATA, $pos, 0;
    my $renderer = build_mt( $t );
    my $rss;
    eval { $rss = $renderer->( $args->{base}, $args->{playlist} )->as_string };
    return $rss;
}

1;

__DATA__
? my ($base, $playlist) = @_;
?= Text::MicroTemplate::encoded_string '<?xml version="1.0" encoding="UTF-8"?>'
<rss version="2.0" xmlns:media="http://search.yahoo.com/mrss" xmlns:itunes="http://www.itunes.com/dtds/podcast-1.0.dtd">
  <channel>
    <title><?= $playlist->{title} ?> - ListPod on <?= $base ?></title>
    <link><?= $base ?>playlist/<?= $playlist->{id} ?></link>
    <description>Podcast generated by ListPod</description>
? for my $video ( @{ $playlist->{videos} } ) {
    <item>
      <link><?= $base ?>playlist/<?= $playlist->{id} ?>#<?= $video->{id} ?></link>
      <title><?= $video->{title} ?></title>
      <enclosure url="<?= $base ?>video/<?= $video->{id} ?>.mp4" type="video/mp4" />
      <media:content url="<?= $base ?>video/<?= $video->{id} ?>.mp4" type="video/mp4"/>
    </item>
? }
  </channel>
</rss>


lib/ListPod/App/Lite/View.pm
package ListPod::App::Lite::View;
use strict;
use warnings;
use Text::MicroTemplate qw(build_mt);

sub render {
    my ($self,$args) = @_;
    my $pos = tell DATA;
    local $/; my $t = <DATA>;
    seek DATA, $pos, 0;
    my $renderer = build_mt( $t );
    my $html;
    eval { $html = $renderer->($args->{base})->as_string };
    return $html;
}

1;

__DATA__
? my ( $base ) = @_;
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
<title>ListPod on <?= $base ?></title>
<link rel="stylesheet" href="http://yusukebe.github.com/ListPod-App-Lite/static/blueprint/screen.css" type="text/css" media="screen, projection" />
<link rel="stylesheet" href="http://yusukebe.github.com/ListPod-App-Lite/static/blueprint/print.css" type="text/css" media="print" />
<!--[if lt IE 8]><link rel="stylesheet" href="http://yusukebe.github.com/ListPod-App-Lite/static/blueprint/ie.css" type="text/css" media="screen, projection" /><![endif]-->
</head>
<body>
<div class="container">
<hr class="space" />
<h1>ListPod on <?= $base ?></h1>
<form action="<?= $base ?>playlist">
<p>
<label>Input <a href="http://www.youtube.com/" target="_blank">YouTube</a>
 Playlist URL</label><br />
<input type="text" name="url" class="title" style="width:800px" />
<input type="submit" value="Submit" style="width:100px;" class="title"/>
</p>
</form>
<br />
<br />
<hr />
<address>listpod-app-lite is developed on
<a href="http://github.com/yusukebe/ListPod-App-Lite" target="_blank">
GitHub
</a> by Yusuke Wada.
This is branche of <a href="http://listpod.tv/">listpod.tv</a>
</address>
</div>
</body>
</html>


lib/ListPod/App/Lite/YouTube.pm
package ListPod::App::Lite::YouTube;
use strict;
use warnings;
use Carp ();
use Encode;

sub new {
    my ( $class, %opt ) = @_;
    Carp::croak("ua parameter is required!") unless $opt{ua};
    my $self = bless { ua => $opt{ua} }, $class;
    $self;
}

sub mp4_url {
    my ($self, $video_id) = @_;
    my $fmt = '18'; #mp4 default
    my $token = $self->token($video_id);
    my $ua = $self->{ua};
    return unless $token;
    my $url =
        "http://www.youtube.com/get_video?video_id=$video_id&t=$token&fmt=$fmt";
    $ua->max_redirect(0);
    my $res = $ua->get( $url );
    return $res->headers->header('Location');
}

sub token {
    my ($self, $video_id) = @_;
    my $url = "http://www.youtube.com/watch?v=$video_id";
    $self->{ua}->get($url);
    $url = "http://www.youtube.com/get_video_info?video_id=$video_id";
    my $res     = $self->{ua}->get($url);
    my $content = $res->content;
    if ( $content =~ m!&token=([^&]+)&! ) {
        return $1;
    }
    else {
        #XXX TODO: Check status code.
        warn $res->status_line . "\n";
        return;
    }
}

sub playlist {
    my ( $self, $playlist_id ) = @_;
    my $url = "http://gdata.youtube.com/feeds/api/playlists/$playlist_id";
    my $res = $self->{ua}->get( $url );
    my $content = $res->content;
    my ($title) = $content =~ m!>([^<>]+)</title><subtitle!m;
    my $videos;
    while ( $content =~ m!vi/([^/]+)/0.jpg'.*?<media:title type='plain'>([^<>]+)</media:title>!g ){
        push @$videos, { title => $2, id => $1 };
    }
    return { videos => $videos, title => $title, id => $playlist_id };
}

1;
__END__

=head1 NAME

ListPod::App::Lite::YouTube - Utilities for Enjoying YouTube Hacking.

=head1 SYNOPSIS

  my $youtube = ListPod::App::Lite::YouTube->new( ua => $ua );
  my $token = $youtube->token( $video_id );
  my $mp4_url = $youtube->mp4_url( $video_id );
  my $playlist = $youtube->playlist( $playlist_id );

=head1 DESCRIPTION

Enjoy YouTube Hacking ♥

=head1 AUTHOR

Yusuke Wada E<lt>yusuke at kamawada.comE<gt>

=head1 SEE ALSO

http://listpod.tv/

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/ListPod/App/Lite/UserAgent.pm
package ListPod::App::Lite::UserAgent;
use strict;
use warnings;
use base ('LWP::UserAgent');

sub new {
    my ( $class, %opt ) = @_;
    my $self = bless {}, $class;
    my $cookie_jar = $self->_cookie_jar($opt{cookie_file}) if $opt{cookie_file};
    $self->cookie_jar( $cookie_jar );
    $self;
}

sub http_headers_raw {
    my $self = shift;
    my $jar  = $self->cookie_jar;
    my $cookie = '';
    if ($jar) {
        my $array = [];
        $jar->scan(
            sub {
                return unless $_[4] =~ /\.youtube\.com$/;
                push @$array, "$_[1]=$_[2]";
            }
        );
        $cookie = join "; ", @$array;
    }
    my $headers = {
        'cookie' => $cookie,
        #XXX
    };
    return $headers;
}

sub _cookie_jar {
    my ( $self, $file ) = @_;
    return {} unless $file;
    if( $file =~ /\.plist$/ ) {
        return $self->_safari_cookie_jar( $file );
    }elsif( $file =~ /\.sqlite$/ ){
        require HTTP::Cookies::Mozilla;
        HTTP::Cookies::Mozilla->new( file => $file );
    }else{
        require HTTP::Cookies;
        HTTP::Cookies->new( file => $file );
    }
}

sub _safari_cookie_jar {
    my ( $self, $file ) = @_;
    # parse Safari cookie, it can't work with HTTP::Cookies::Safari
    # maybe problem of "Expires"??
    require HTTP::Cookies;
    require Mac::PropertyList;
    my $cookie_jar = HTTP::Cookies->new;
    open my ($fh), $file or return;
    my $data    = do { local $/; <$fh> };
    my $plist   = Mac::PropertyList::parse_plist($data);
    my $cookies = $plist->value;
    for my $hash (@$cookies) {
        my $cookie = $hash->value;
        if ( $cookie->{Domain}->value =~ /\.youtube\.com/ ) {
            $cookie_jar->set_cookie(
                undef,
                $cookie->{Name}->value,
                $cookie->{Value}->value,
                $cookie->{Path}->value,
                $cookie->{Domain}->value
            );
        }
    }
    close $fh;
    return $cookie_jar;
}

1;
__END__

=head1 NAME

ListPod::App::Lite::UserAgent - UserAgent for fight! It's LWP::UserAgent based.

=head1 SYNOPSIS

  my $ua = ListPod::App::Lite::UserAgent->new(%options);
  my $res = $ua->get( 'http://www.youtube.com/' );

=head1 DESCRIPTION

It's subclass of LWP::UserAgent.

=head1 AUTHOR

Yusuke Wada E<lt>yusuke at kamawada.comE<gt>

=head1 SEE ALSO

http://listpod.tv/

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/ListPod/App/Lite.pm
package ListPod::App::Lite;
use strict;
use warnings;
our $VERSION = '0.01';
use Plack::Request;
use ListPod::App::Lite::View;
use ListPod::App::Lite::RSS;
use ListPod::App::Lite::YouTube;
use ListPod::App::Lite::UserAgent;

sub new {
    my ( $class, %opt ) = @_;
    my $self = bless {
        cookie_file => $opt{cookie_file},
    }, $class;
    $self->{ua} = ListPod::App::Lite::UserAgent->new( cookie_file => $opt{cookie_file} );
    $self->{youtube} = ListPod::App::Lite::YouTube->new( ua => $self->{ua} );
    $self;
}

sub handler {
    my $self = shift;
    sub {
        my $env = shift;
        my $req = Plack::Request->new($env);
        $self->{base} = $req->base unless $self->{base};
        if ( $req->path_info eq '/' ) {
            my $html = ListPod::App::Lite::View->render({ base => $self->{base} });
            return $self->response( $html, 'text/html' );
        }
        elsif( $req->path_info eq "/playlist" ){
            my ($playlist_id) = $req->param('url') =~ /p=([^&\s]+)/;
            return $self->handle_404 unless $playlist_id;
            $self->redirect( $self->{base} . 'playlist/' . $playlist_id );
        }
        elsif( $req->path_info =~ m!/playlist/([0-9A-Z]+)! ){
            my $playlist_id = $1;
            my $playlist = $self->{youtube}->playlist($playlist_id);
            my $rss      = ListPod::App::Lite::RSS->render(
                { base => $self->{base}, playlist => $playlist } );
            return $self->response( $rss, 'application/rss+xml' );
        }
        elsif ( $req->path_info =~ m!/video/([^\.]+)\.mp4! ) {
            my $video_id = $1;
            my $mp4_url  = $self->{youtube}->mp4_url($video_id);
            return [402, ['Content-Type'=>'text/plain'], ['402 Payment Required'] ] unless $mp4_url;
            return $self->handle_video_anyevent($mp4_url) if $self->{cookie_file};
            return $self->redirect($mp4_url);
        }
        else {
            return [ 404, [], ['404 Not Found'] ];
        }
    };
}

sub response {
    my ( $self, $content, $type ) = @_;
    return [ 200, [ 'Content-Type' => $type ], [$content] ];
}

sub handle_404 {
    return [ 404, [], ['404 Not Found'] ];
}

sub redirect {
    my ( $self, $url ) = @_;
    return [ 302, [ Location => $url ], [] ];
}


sub handle_video_anyevent {
    my ($self, $url) = @_;
    warn "Handle video with AnyEvent::HTTP\n";
    require AnyEvent;
    require AnyEvent::HTTP;
    my $headers = $self->{ua}->http_headers_raw;
    my $cv      = AnyEvent->condvar;
    warn "Donwloading ... $url\n";
    AnyEvent::HTTP::http_get(
        $url,
        headers => $headers,
        sub {
            my $content = shift;
            $cv->send(
                [
                    200,
                    [
                        'Content-Type'   => 'video/mp4',
                        'Content-Length' => length $content
                    ],
                    [$content]
                ]
            );
        }
    );
    return sub {
        my $start_response = shift;
        $cv->cb(
            sub {
                warn "Downloaded\n";
                $start_response->( shift->recv );
            }
        );
    };
}



1;
__END__

=head1 NAME

ListPod::App::Lite - Modules for listpod-app-lite command.

=head1 SYNOPSIS

in your .psgi

  use ListPod::App::Lite;
  my $app = ListPod::App::Lite->new(%options);
  return $app->handler;

=head1 DESCRIPTION

ListPod::App::Lite is Module(s) for listpod-app-lite command.

=head1 AUTHOR

Yusuke Wada E<lt>yusuke at kamawada.comE<gt>

=head1 SEE ALSO

http://listpod.tv/

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
